search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+                                   "description", g_variant_new_string (data->desc));
+        } else {
+            g_variant_builder_add (&meta, "{sv}",
+                                   "name", NULL);
+            g_variant_builder_add (&meta, "{sv}",
+                                   "gicon", NULL);
+            g_variant_builder_add (&meta, "{sv}",
+                                   "description", NULL);
+        }
+        g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+/* FIXME: Search was copied from yelp-search-entry and slightly adapted.
+ * When search in yelp gets adjusted we should export it and use it here.
+ */
+static gboolean
+match_word (const gchar *search_phrase,
+            const gchar *title_casefold,
+            const gchar *desc_casefold)
+{
+    static GRegex *nonword = NULL;
+    gint stri;
+    gchar **strs;
+    gboolean ret = TRUE;
+
+    if (nonword == NULL)
+        nonword = g_regex_new ("\\W", 0, 0, NULL);
+    g_assert (nonword != NULL);
+
+    strs = g_regex_split (nonword, search_phrase, 0);
+    for (stri = 0; strs[stri]; stri++) {
+        if (!title_casefold || !strstr (title_casefold, strs[stri])) {
+            if (!desc_casefold || !strstr (desc_casefold, strs[stri])) {
+                ret = FALSE;
+                break;
+            }
+        }
+    }
+
+    g_strfreev (strs);
+
+    return ret;
+}
+
+static GVariant *
+get_search_results (gchar      **terms,
+                    GHashTable  *page_data)
+{
+    GVariantBuilder results;
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    const gchar *page_id;
+    PageData *data;
+
+    g_hash_table_iter_init (&iter, page_data);
+    g_variant_builder_init (&results, G_VARIANT_TYPE ("as"));
+
+    while (g_hash_table_iter_next (&iter, (gpointer *) &page_id, (gpointer *) &data)) {
+        if (match_word (term, data->title_casefold, data->desc_casefold))
+            g_variant_builder_add (&results, "s", page_id);
+    }
+
+    g_free (term);
+
+    return g_variant_new ("(as)", &results);
+}
+
+typedef GVariant * (*ResultGetter) (gchar      **terms,
+                                    GHashTable  *page_data);
+
+typedef struct
+{
+    YelpSearchProviderApp  *app;
+    GDBusMethodInvocation  *invocation;
+    gchar                 **terms;
+    ResultGetter            result_getter;
+} DelayedResultGetter;
+
+static gboolean
+check_page_data (gpointer user_data)
+{
+    DelayedResultGetter* delayed = user_data;
+
+    if (g_hash_table_size (delayed->app->page_data_hash_map) > 0) {
+        g_dbus_method_invocation_return_value (delayed->invocation,
+                                               delayed->result_getter (delayed->terms,
+                                                                       delayed->app->page_data_hash_map));
+        return G_SOURCE_REMOVE;
+    }
+
+    return G_SOURCE_CONTINUE;
```

Comment:
Instead of always re-trying, I think that a better approach would be for the delayed object to directly call back when the hash map is populated.